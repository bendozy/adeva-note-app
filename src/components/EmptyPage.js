import React from 'react';
import PropTypes from 'prop-types';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(theme => ({
  body: {
    marginTop: theme.spacing(5),
  },
}));

const EmptyPage = ({ title, text }) => {
  const classes = useStyles();

  return (
    <Grid>
      <Grid>
        <Typography variant="h4" color="inherit" noWrap>
          {title}
        </Typography>
      </Grid>
      <Grid>
        <Typography
          variant="body1"
          component="p"
          color="inherit"
          className={classes.body}
        >
          {text}
        </Typography>
      </Grid>
    </Grid>
  );
};

EmptyPage.defaultProps = {
  title: 'Page Not Found',
  text: 'The requested url could not be found',
};

EmptyPage.propTypes = {
  title: PropTypes.string,
  text: PropTypes.string,
};

export default EmptyPage;
