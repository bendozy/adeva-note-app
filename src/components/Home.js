import React from 'react';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(theme => ({
  body: {
    marginTop: theme.spacing(5),
  },
}));

const Home = () => {
  const classes = useStyles();

  return (
    <Grid>
      <Grid>
        <Typography variant="h4" color="inherit" noWrap>
          Welcome
        </Typography>
      </Grid>
      <Grid>
        <Typography
          variant="body1"
          component="p"
          color="inherit"
          className={classes.body}
        >
          Hello there, thanks you for checking out this simple notes application.
          With this app, you can create, update, delete and search for notes.
        </Typography>
        <Typography
          variant="body1"
          component="p"
          color="inherit"
          className={classes.body}
        >
          This appliction uses localStorage for persistence. To get startted,
          click on the &quot;Add New Note&quot; button on the sidebar.
        </Typography>
      </Grid>
    </Grid>
  );
};

export default Home;
