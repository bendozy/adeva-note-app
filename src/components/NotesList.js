import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { makeStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import InputBase from '@material-ui/core/InputBase';
import Grid from '@material-ui/core/Grid';
import IconButton from '@material-ui/core/IconButton';
import Button from '@material-ui/core/Button';
import MenuIcon from '@material-ui/icons/Menu';
import SearchIcon from '@material-ui/icons/Search';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';

const useStyles = makeStyles({
  search: {
    marginTop: 5,
    display: 'flex',
    alignItems: 'center',
    width: '100%',
  },
  list: {
    // height: '93vh',
  },
  input: {
    marginLeft: 8,
    flex: 1,
  },
  iconButton: {
    padding: 10,
  },
});

const NotesList = ({ notes }) => {
  const classes = useStyles();
  const [searchText, setSearchText] = useState('');

  return (
    <Grid>
      <Grid>
        <Paper className={classes.search}>
          <IconButton className={classes.iconButton} aria-label="Menu">
            <MenuIcon />
          </IconButton>
          <InputBase
            className={classes.input}
            placeholder="Search Notes"
            inputProps={{ 'aria-label': 'Search Notes' }}
            value={searchText}
            onChange={event => setSearchText(event.target.value)}
          />
          <IconButton className={classes.iconButton} aria-label="Search">
            <SearchIcon />
          </IconButton>
        </Paper>
      </Grid>
      <Grid>
        <List component="nav" aria-label="Secondary mailbox folders">
          {notes
            .filter(note =>
              note.title.toLowerCase().includes(searchText.toLowerCase())
                || note.details.toLowerCase().includes(searchText.toLowerCase()))
            .map((note, index) => (
              <ListItem
                key={note.id}
                button
                divider={notes.length !== index + 1}
                component={Link}
                to={`/notes/${note.id}`}
              >
                <ListItemText primary={note.title} />
              </ListItem>
            ))
          }
        </List>
      </Grid>
      <Grid>
        <Button
          fullWidth
          variant="contained"
          color="primary"
          component={Link}
          to="/notes/add"
        >
          Add New Note
        </Button>
      </Grid>
    </Grid>
  );
};

NotesList.defaultProps = {
  notes: [],
};

NotesList.propTypes = {
  notes: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    details: PropTypes.string.isRequired,
  })),
};

const mapStateToProps = ({ notes }) => ({ notes });

export default connect(mapStateToProps)(NotesList);
