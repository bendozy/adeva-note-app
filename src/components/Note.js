import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Link } from 'react-router-dom';
import classNames from 'classnames';
import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core/styles';
import Fab from '@material-ui/core/Fab';
import Icon from '@material-ui/core/Icon';
import DeleteIcon from '@material-ui/icons/Delete';
import Typography from '@material-ui/core/Typography';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import Button from '@material-ui/core/Button';
import Snackbar from '@material-ui/core/Snackbar';
import EmptyPage from './EmptyPage';
import { deleteNote as deleteNoteAction } from '../actions/notes';

const useStyles = makeStyles(theme => ({
  details: {
    marginTop: theme.spacing(5),
  },
  fab: {
    margin: theme.spacing(1),
    position: 'fixed',
    bottom: theme.spacing(2),
    right: theme.spacing(2),
  },
  fabEdit: {
    right: theme.spacing(10),
  },
  fabDelete: {
    right: theme.spacing(2),
  },
}));

const Note = ({
  history,
  deleteNoteAction: deleteNote,
  note,
  note: {
    id,
    title,
    details,
  } = {},
}) => {
  const classes = useStyles();
  const [open, setOpen] = useState(false);
  const [snackOpen, setSnackOpen] = useState(false);

  const handleOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  const handleSubmit = () => {
    deleteNote(note.id);
    setOpen(false);
    setSnackOpen(true);
    history.push('/');
  };

  if (!note) {
    return (
      <EmptyPage
        title="Note Not Found"
        text="The requested note is not available"
      />
    );
  }

  return (
    <React.Fragment>
      <Grid>
        <Grid>
          <Typography variant="h4" color="inherit" noWrap>
            {title}
          </Typography>
        </Grid>
        <Grid>
          <Typography
            variant="body1"
            component="p"
            color="inherit"
            className={classes.details}
          >
            {details}
          </Typography>
        </Grid>
      </Grid>
      <Fab
        color="primary"
        aria-label="Edit Note"
        className={classNames(classes.fab, classes.fabEdit)}
        component={Link}
        to={`/notes/${id}/edit`}
      >
        <Icon>edit_icon</Icon>
      </Fab>
      <Fab
        color="secondary"
        aria-label="Delete Note"
        className={classNames(classes.fab, classes.fabDelete)}
        onClick={handleOpen}
      >
        <DeleteIcon />
      </Fab>
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title">Delete Note?</DialogTitle>
        <DialogContent>
          <DialogContentText id="alert-dialog-description">
            Are you sure you want to delete this note.
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose} color="primary">
            Cancel
          </Button>
          <Button onClick={handleSubmit} color="secondary" autoFocus>
            Yes, Delete
          </Button>
        </DialogActions>
      </Dialog>
      <Snackbar
        anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
        open={snackOpen}
        onClose={() => setSnackOpen(false)}
        autoHideDuration={4000}
        ContentProps={{
          'aria-describedby': 'message-id',
        }}
        message={<span id="message-id">Note deleted successfully</span>}
      />
    </React.Fragment>
  );
};

Note.defaultProps = {
  note: undefined,
};

Note.propTypes = {
  note: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    details: PropTypes.string.isRequired,
  }),
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  deleteNoteAction: PropTypes.func.isRequired,
};

const mapStateToProps = ({ notes }, { match: { params } }) =>
  ({
    note: notes.find(note => note.id === parseInt(params.noteId, 10)),
  });

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      deleteNoteAction,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(Note);
