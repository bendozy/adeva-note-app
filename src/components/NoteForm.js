import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import EmptyPage from './EmptyPage';
import {
  createNote as createNoteAction,
  updateNote as updateNoteAction,
} from '../actions/notes';

const useStyles = makeStyles(theme => ({
  form: {
    width: '100%',
    marginTop: theme.spacing(1),
  },
}));

const NoteForm = ({
  note,
  nextId,
  history,
  isEditMode,
  createNoteAction: createNote,
  updateNoteAction: updateNote,
}) => {
  const classes = useStyles();
  const [values, setValues] = useState({
    title: note ? note.title : '',
    details: note ? note.details : '',
  });

  const handleChange = name => (event) => {
    setValues({ ...values, [name]: event.target.value });
  };

  const formatInput = name => (event) => {
    setValues({ ...values, [name]: event.target.value.trim() });
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    if (note) {
      updateNote(Object.assign({}, note, values));
      history.push(`/notes/${note.id}`);
    } else {
      createNote(values);
      history.push(`/notes/${nextId}`);
    }
  };

  if (isEditMode && !note) {
    return (
      <EmptyPage
        title="Note Not Found"
        text="The requested note is not available"
      />
    );
  }

  return (
    <Grid>
      <Grid>
        <Typography variant="h4" color="inherit" noWrap>
          {note ? 'Update' : 'New'} Note
        </Typography>
      </Grid>
      <form
        className={classes.form}
        onSubmit={handleSubmit}
      >
        <Grid>
          <TextField
            required
            id="title"
            label="Note Title"
            value={values.title}
            fullWidth
            onChange={handleChange('title')}
            onBlur={formatInput('title')}
            margin="normal"
          />
        </Grid>
        <Grid>
          <TextField
            required
            id="title"
            label="Note Details"
            value={values.details}
            fullWidth
            onChange={handleChange('details')}
            multiline
            rows={20}
            margin="normal"
          />
        </Grid>
        <Grid>
          <Button
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
          >
            {note ? 'Update' : 'Create'} Note
          </Button>
        </Grid>
      </form>
    </Grid>
  );
};

NoteForm.defaultProps = {
  note: undefined,
};

NoteForm.propTypes = {
  note: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    details: PropTypes.string.isRequired,
  }),
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  nextId: PropTypes.number.isRequired,
  isEditMode: PropTypes.bool.isRequired,
  createNoteAction: PropTypes.func.isRequired,
  updateNoteAction: PropTypes.func.isRequired,
};

const mapStateToProps = ({ notes, nextId }, { match: { params } }) =>
  ({
    note: notes.find(note => note.id === parseInt(params.noteId, 10)),
    isEditMode: !!params.noteId,
    nextId,
  });

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      createNoteAction,
      updateNoteAction,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(NoteForm);
