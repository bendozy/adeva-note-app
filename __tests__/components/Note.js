import React from 'react';
import toJson from 'enzyme-to-json';
import Typography from '@material-ui/core/Typography';
import Fab from '@material-ui/core/Fab';
import Dialog from '@material-ui/core/Dialog';
import { Note, mapStateToProps, mapDispatchToProps } from '../../src/components/Note';
import EmptyPage from '../../src/components/EmptyPage';

describe('<Note />', () => {
  it('should render a <Note /> component', () => {
    const wrapper = shallow(<Note
      deleteNoteAction={jest.fn()}
      history={{
        push: jest.fn(),
      }}
      note={{
        id: 1,
        title: 'Test Example',
        details: 'Testing detail',
      }}
    />);


    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(Typography)).toHaveLength(2);
    expect(wrapper.find(Typography).first().text()).toBe('Test Example');
    expect(wrapper.find(Typography).at(1).text()).toBe('Testing detail');
    expect(wrapper.find(Fab)).toHaveLength(2);
    expect(wrapper.find(Dialog)).toHaveLength(1);
  });

  it('should show a dialog if the delete button is clicked', () => {
    const wrapper = shallow(<Note
      deleteNoteAction={jest.fn()}
      history={{
        push: jest.fn(),
      }}
      note={{
        id: 1,
        title: 'Test Example',
        details: 'Testing detail',
      }}
    />);

    wrapper.find(Fab).at(0).simulate('click');
  });

  it('should render a <EmptyPage /> component when no note is found', () => {
    const wrapper = shallow(<Note
      deleteNoteAction={jest.fn()}
      history={{
        push: jest.fn(),
      }}
    />);

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(EmptyPage)).toHaveLength(1);
  });

  it('check for the correct note title', () => {
    expect(mapStateToProps({
      notes: [{
        id: 1,
        title: 'Test Example',
        details: 'Testing detail',
      }],
    }, {
      match: {
        params: {
          noteId: 1,
        },
      },
    }).note.title).toBe('Test Example');
  });

  it('should discpatch searchMovie action', () => {
    const dispatch = jest.fn();

    mapDispatchToProps(dispatch).deleteNoteAction();
    expect(dispatch.mock.calls[0][0].type).toEqual('DELETE_NOTE');
  });
});
