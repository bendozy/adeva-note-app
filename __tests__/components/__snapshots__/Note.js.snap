// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Note /> should render a <EmptyPage /> component when no note is found 1`] = `
<EmptyPage
  text="The requested note is not available"
  title="Note Not Found"
/>
`;

exports[`<Note /> should render a <Note /> component 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Typography))
        color="inherit"
        noWrap={true}
        variant="h4"
      >
        Test Example
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-details-1"
        color="inherit"
        component="p"
        variant="body1"
      >
        Testing detail
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Fab))
    aria-label="Edit Note"
    className="makeStyles-fab-2 makeStyles-fabEdit-3"
    color="primary"
    component={[Function]}
    to="/notes/1/edit"
  >
    <WithStyles(ForwardRef(Icon))>
      edit_icon
    </WithStyles(ForwardRef(Icon))>
  </WithStyles(ForwardRef(Fab))>
  <WithStyles(ForwardRef(Fab))
    aria-label="Delete Note"
    className="makeStyles-fab-2 makeStyles-fabDelete-4"
    color="secondary"
    onClick={[Function]}
  >
    <DeleteIcon />
  </WithStyles(ForwardRef(Fab))>
  <WithStyles(ForwardRef(Dialog))
    aria-describedby="alert-dialog-description"
    aria-labelledby="alert-dialog-title"
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(DialogTitle))
      id="alert-dialog-title"
    >
      Delete Note?
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))>
      <WithStyles(ForwardRef(DialogContentText))
        id="alert-dialog-description"
      >
        Are you sure you want to delete this note.
      </WithStyles(ForwardRef(DialogContentText))>
    </WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogActions))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        autoFocus={true}
        color="secondary"
        onClick={[Function]}
      >
        Yes, Delete
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(DialogActions))>
  </WithStyles(ForwardRef(Dialog))>
</Fragment>
`;
